### DISCUSSION TITLE
Scaling is not performed because triggers are not active

### DISCUSSION THREAD

**Discussion by slengieza** [kedacore/keda#1387] [Q&A / Need Help]
I am new to Kubernetes and I am trying to manage my standalone Azure Container Instances with Azure Kubernetes Service. I have been able to deploy the application using AKS and I have been following [this sample](https://github.com/kedacore/sample-dotnet-worker-servicebus-queue/blob/main/connection-string-scenario.md) to auto-scale the application based on the number of messages in Azure Service Bus Topics.

When I initially deploy the ScaledObject I can see the deployments scale up correctly (if messages are already added to the Topic) and then they scale back down to 0 when the messages have all been consumed. However, when I add more messages to the Topic or deploy the ScaledObject while there are no messages queued I am not seeing the deployment scale up as it did before. 

When I run `kubectl describe scaledobject.keda.sh/wwml-aks-canary-scaler` I see the following:

```
Status:
  Conditions:
    Message:  ScaledObject is defined correctly and is ready for scaling
    Reason:   ScaledObjectReady
    Status:   True
    Type:     Ready
    Message:  Scaling is not performed because triggers are not active
    Reason:   ScalerNotActive
    Status:   False
    Type:     Active
```

I have been looking online to see if I could figure out why the triggers would not be "Active" but I wasn't able to find anything. Any help would be greatly appreciated. Thanks!

This is my autoscale yaml file:
```apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: trigger-auth-service-bus-canary
spec:
  secretTargetRef:
  - parameter: connection
    name: secrets-management
    key: servicebus-connectionstring
---
apiVersion: v1
kind: Secret
metadata:
  name: secrets-management
  labels:
    app: wwml-aks-canary
data:
  servicebus-connectionstring: <base 64 encoded connection string>
---
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: wwml-aks-canary-scaler
spec:
  scaleTargetRef:
    name: wwml-aks-canary
  pollingInterval: 30
  cooldownPeriod:  300
  minReplicaCount: 0
  maxReplicaCount: 100
  advanced:
    restoreToOriginalReplicaCount: true
    horizontalPodAutoscalerConfig:
      behavior:
        scaleDown:
          stabilizationWindowSeconds: 300
          policies:
          - type: Percent
            value: 100
            periodSeconds: 15
  triggers:
  - type: azure-servicebus
    metadata:
      topicName: canary-topic
      subscriptionName: Highest
    authenticationRef:
      name: trigger-auth-service-bus-canary
  - type: azure-servicebus
    metadata:
      topicName: canary-topic
      subscriptionName: High
    authenticationRef:
      name: trigger-auth-service-bus-canary
  - type: azure-servicebus
    metadata:
      topicName: canary-topic
      subscriptionName: Normal
    authenticationRef:
      name: trigger-auth-service-bus-canary
  - type: azure-servicebus
    metadata:
      topicName: canary-topic
      subscriptionName: Low
    authenticationRef:
      name: trigger-auth-service-bus-canary
  - type: azure-servicebus
    metadata:
      topicName: canary-topic
      subscriptionName: Lowest
    authenticationRef:
      name: trigger-auth-service-bus-canary
```

**Comment by tomkerkhove** [comment]
> However, when I add more messages to the Topic or deploy the ScaledObject while there are no messages queued I am not seeing the deployment scale up as it did before.

When deploying the ScaledObject it will behave the same way as it was before so this will not help, but if messages are being added it should if it matches `messageCount`. In this case, it is not specified so it will be using 5 ([docs](https://keda.sh/docs/2.0/scalers/azure-service-bus/); how many did you configure?

Can you talk about the application that you are deploying because now you add triggers for all your subscriptions and will take all these metrics into account.

I see you configure `advanced` but typically you don't need this unless you really want that control, do you want this?

**Reply by slengieza** [reply]
I added 30 messages to one of the subscriptions so I figured that should trigger some scaling. Also the application is just a docker container that reads messages from a topic and runs Edge using selenium according to those messages. 

I only added the `advanced` section in an effort to get things to work, but will remove it since that doesnt seem to be the issue. 

After doing some more digging I ran `kubectl logs keda-operator-7fcb5c74bc-9gdxs -n keda` and saw this error pertaining to not being able to find some secrets. 

```
2020-12-01T16:20:42.650Z        ERROR   controller      Reconciler error        {"reconcilerGroup": "keda.sh", "reconcilerKind": "ScaledObject", "controller": "scaledobject", "name": "wwml-aks-canary-scaler", "namespace": "default", "error": "error resolving secrets for ScaleTarget: error resolving secret name &SecretKeySelector{LocalObjectReference:LocalObjectReference{Name:wwml-secrets,},Key:hit-app-service-key,Optional:*true,} for env HIT_APP_SVC_KEY in namespace default"}
```

I am using a secret provided to fetch secrets from Azure Key Vault, this is my yaml file:

```
apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
kind: SecretProviderClass
metadata:
  name: wwml-key-vault
spec:
  provider: azure
  secretObjects:
  - secretName: wwml-secrets
    type: Opaque
    labels:                                   
      environment: "prod"
    data: 
    - objectName: HIT-APP-SERVICE-KEY
      key: hit-app-service-key
    - objectName: WWML-SERVICE-BUS-KEY
      key: wwml-service-bus-key
    - objectName: WWML-SERVICE-BUS-CONNECTION-STRING
      key: wwml-service-bus-connection-string
  parameters:
    usePodIdentity: "false"
    userAssignedIdentityID: "<user assigned identity id>"
    keyvaultName: "wwml-key-vault"
    objects:  |
      array:
        - |
          objectName: HIT-APP-SERVICE-KEY
          objectType: secret
        - |
          objectName: WWML-SERVICE-BUS-KEY
          objectType: secret
        - |
          objectName: WWML-SERVICE-BUS-CONNECTION-STRING
          objectType: secret
    resourceGroup: "wwml-wu2"
    subscriptionId: "<subscription id>"
    tenantId: "<tenant id>"
```

and then this is my deployment yaml file:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: wwml-aks-canary
  labels:
    aadpodidbinding: "edge-wwml-sp"
spec:
  selector:
    matchLabels:
      app: wwml-aks-canary
  template:
    metadata:
      labels:
        app: wwml-aks-canary
    spec:
      nodeSelector:
        "beta.kubernetes.io/os": "windows"
      containers:
        - name: wwml-aks-canary
          image: dsesregistry.azurecr.io/ml_features/wwml/canary:latest
          resources:
            requests:
              cpu: "1"
              memory: 4Gi
            limits:
              cpu: "2"
              memory: 8Gi
          env:
          - name: HIT_APP_SVC_KEY
            valueFrom:
              secretKeyRef:
                name: wwml-secrets
                key: hit-app-service-key
          - name: WWML_SERVICE_BUS_KEY
            valueFrom:
              secretKeyRef:
                name: wwml-secrets
                key: wwml-service-bus-key
          - name: WWML_SERVICE_BUS_CONNECTION_STRING
            valueFrom:
                secretKeyRef:
                  name: wwml-secrets
                  key: wwml-service-bus-connection-string
          volumeMounts:
          - name: secrets-store-inline
            mountPath: "/mnt/secrets-store"
            readOnly: true
      volumes:
        - name: secrets-store-inline
          csi:
            driver: secrets-store.csi.k8s.io
            readOnly: true
            volumeAttributes:
              secretProviderClass: "wwml-key-vault"
            nodePublishSecretRef:
              name: secrets-store-creds
```
When I have an instance of the deployment running I can see the `wwml-secrets` pod running but when Keda scales the deployment down to 0 I see the secrets pod go away so I am assuming that I need to persist the `SecretProviderClass`, is that right?

**Reply by tomkerkhove** [reply]
I looks like it's looking for `HIT_APP_SVC_KEY` instead of `WWML_SERVICE_BUS_CONNECTION_STRING`, no?

Can you paste all your resources again please? Because this is not inline with what you've initially reported or is it?

**Reply by slengieza** [reply]
The deployment uses `HIT_APP_SVC_KEY` and `WWML_SERVICE_BUS_CONNECTION_STRING` which are in the `SecretProviderClass` but the `ScaledObject` is supposed to just be using inline base64 encoded connection string. 

Here are the three files I have deployed (they are what I posted above):

Yaml file for the actual application:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: wwml-aks-canary
  labels:
    aadpodidbinding: "edge-wwml-sp"
spec:
  selector:
    matchLabels:
      app: wwml-aks-canary
  template:
    metadata:
      labels:
        app: wwml-aks-canary
    spec:
      nodeSelector:
        "beta.kubernetes.io/os": "windows"
      containers:
        - name: wwml-aks-canary
          image: dsesregistry.azurecr.io/ml_features/wwml/canary:latest
          resources:
            requests:
              cpu: "1"
              memory: 4Gi
            limits:
              cpu: "2"
              memory: 8Gi
          env:
          - name: HIT_APP_SVC_KEY
            valueFrom:
              secretKeyRef:
                name: wwml-secrets
                key: hit-app-service-key
          - name: WWML_SERVICE_BUS_KEY
            valueFrom:
              secretKeyRef:
                name: wwml-secrets
                key: wwml-service-bus-key
          - name: WWML_SERVICE_BUS_CONNECTION_STRING
            valueFrom:
                secretKeyRef:
                  name: wwml-secrets
                  key: wwml-service-bus-connection-string
          volumeMounts:
          - name: secrets-store-inline
            mountPath: "/mnt/secrets-store"
            readOnly: true
      volumes:
        - name: secrets-store-inline
          csi:
            driver: secrets-store.csi.k8s.io
            readOnly: true
            volumeAttributes:
              secretProviderClass: "wwml-key-vault"
            nodePublishSecretRef:
              name: secrets-store-creds
```

Yaml file for providing secrets to the deployment above:
```
apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
kind: SecretProviderClass
metadata:
  name: wwml-key-vault
spec:
  provider: azure
  secretObjects:
  - secretName: wwml-secrets
    type: Opaque
    labels:                                   
      environment: "prod"
    data: 
    - objectName: HIT-APP-SERVICE-KEY
      key: hit-app-service-key
    - objectName: WWML-SERVICE-BUS-KEY
      key: wwml-service-bus-key
    - objectName: WWML-SERVICE-BUS-CONNECTION-STRING
      key: wwml-service-bus-connection-string
  parameters:
    usePodIdentity: "false"
    userAssignedIdentityID: <user assigned identity id>
    keyvaultName: "wwml-key-vault"
    objects:  |
      array:
        - |
          objectName: HIT-APP-SERVICE-KEY
          objectType: secret
        - |
          objectName: WWML-SERVICE-BUS-KEY
          objectType: secret
        - |
          objectName: WWML-SERVICE-BUS-CONNECTION-STRING
          objectType: secret
    resourceGroup: "wwml-wu2"
    subscriptionId: <subscription id>
    tenantId: <tenant id>
```

Yaml file for autoscaling:
```
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: trigger-auth-service-bus-canary
spec:
  secretTargetRef:
  - parameter: connection
    name: secrets-management
    key: servicebus-connectionstring
---
apiVersion: v1
kind: Secret
metadata:
  name: secrets-management
  labels:
    app: wwml-aks-canary
data:
  servicebus-connectionstring: <base64 encoded connection string>
---
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: wwml-aks-canary-scaler
spec:
  scaleTargetRef:
    name: wwml-aks-canary
  pollingInterval: 30
  cooldownPeriod:  300
  minReplicaCount: 0
  maxReplicaCount: 100
  triggers:
  - type: azure-servicebus
    metadata:
      topicName: canary-topic
      subscriptionName: Highest
    authenticationRef:
      name: trigger-auth-service-bus-canary
  - type: azure-servicebus
    metadata:
      topicName: canary-topic
      subscriptionName: High
    authenticationRef:z
      name: trigger-auth-service-bus-canary
  - type: azure-servicebus
    metadata:
      topicName: canary-topic
      subscriptionName: Normal
    authenticationRef:
      name: trigger-auth-service-bus-canary
  - type: azure-servicebus
    metadata:
      topicName: canary-topic
      subscriptionName: Low
    authenticationRef:
      name: trigger-auth-service-bus-canary
  - type: azure-servicebus
    metadata:
      topicName: canary-topic
      subscriptionName: Lowest
    authenticationRef:
      name: trigger-auth-service-bus-canary
```

Basically the Deplyoment uses the `SecretProviderClass` for environment variables and the connection to the service bus but the `ScaledObject` uses the `Secret` defined in its own yaml file (this is only because I wanted to get it working before switching the auto scale file to use the `SecretProviderClass`).

**Reply by slengieza** [reply]
When I change `minReplicaCount` to 1 everything works as expected. I assume this is because the deployment never goes away and therefore the `wml-secrets` class still exists so the `ScaledObject` is able to find all of the secrets.

**Reply by tomkerkhove** [reply]
I'm not sure if that is correct, because from what I'm reading and is defined KEDA will not require anything from the deployment and purely relies on the Secret that you are providing.

I would say, can you provide some more logs from KEDA because the ones provided above don't make sense because it's referring to different secret & environment variables.

Now if you would rely on the `SecretProviderClass` which is mounted to your app, then I think scale to 0 will not be possible indeed. That said, we don't support mounting secrets in KEDA as well so that would be a limitation /cc @zroubalik 

*PS - I saw you are using pod identity, don't forget that you can also enable this with KEDA.*

**Reply by slengieza** [reply]
I am using a `SecretProviderClass` to mount secrets. Is that something that will be supported in the future? 

My workaround for now is having a small container constantly running that also mounts those secrets so that the secret Pod is present when KEDA is trying to scale from 0-1.

**Reply by tomkerkhove** [reply]
It would be good if you can open an issue for this; that would feel like a new secret provider for `TriggerAuthentication`, what do you think @zroubalik?

**Reply by zroubalik** [reply]
Yeah agree with Tom, open an issue if you are interested in this secret provider, `SecretProviderClass` is unfortunately not supported currently.

**Reply by slengieza** [reply]
Okay will do, thanks

**Comment by slengieza** [comment]
Created Issue here: https://github.com/kedacore/keda/issues/1407

**Comment by jfborie** [comment]
In case you have this error go into the keda namespace and look at the log in the keda-operator pod, you might get more useful information.

**Comment by RedEye-Developers** [comment]
`Scaling is not performed because triggers are not active`
 
have same issue any solution ?

**Comment by RedEye-Developers** [comment]
```
Status:
  Authentications Types:  authentication-manager-consumer
  Conditions:
    Message:  ScaledObject is defined correctly and is ready for scaling
    Reason:   ScaledObjectReady
    Status:   True
    Type:     Ready
    Message:  Scaling is not performed because triggers are not active
    Reason:   ScalerNotActive
    Status:   False
    Type:     Active
    Message:  No fallbacks are active on this scaled object
    Reason:   NoFallbackFound
    Status:   False
    Type:     Fallback
    Status:   Unknown
    Type:     Paused
```
